tests/python/unittest/test_custom_datatypes.py

```diff
@@ -0,0 +1,407 @@
+                    groups=32,
+                    kernel_size=(3, 3),
+                    fref=lambda x, w: topi.testing.
+                    depthwise_conv2d_python_nchw(x, w, (1, 1), "SAME"))
+
+    # CUDA is disabled for 'direct' schedule:
+    # https://github.com/dmlc/tvm/pull/3070#issuecomment-486597553
+    # group conv2d
+    dshape = (1, 32, 18, 18)
+    kshape = (32, 4, 3, 3)
+    run_test_conv2d(src_dtype,
+                    dst_dtype,
+                    1,
+                    dshape,
+                    kshape,
+                    padding=(1, 1),
+                    channels=32,
+                    groups=8,
+                    kernel_size=(3, 3),
+                    except_targets=['cuda'])
+    # also group conv2d
+    dshape = (1, 32, 18, 18)
+    kshape = (64, 1, 3, 3)
+    run_test_conv2d(src_dtype,
+                    dst_dtype,
+                    1,
+                    dshape,
+                    kshape,
+                    padding=(1, 1),
+                    channels=64,
+                    groups=32,
+                    kernel_size=(3, 3),
+                    except_targets=['cuda'])
+
+    # normal conv2d
+    dshape = (1, 3, 224, 224)
+    kshape = (10, 3, 3, 3)
+    run_test_conv2d(src_dtype,
+                    dst_dtype,
+                    1,
+                    dshape,
+                    kshape,
+                    padding=(1, 1),
+                    channels=10,
+                    kernel_size=(3, 3))
+
+    # dilated conv2d
+    dshape = (1, 3, 18, 18)
+    kshape = (10, 3, 3, 3)
+    run_test_conv2d(src_dtype,
+                    dst_dtype,
+                    1,
+                    dshape,
+                    kshape,
+                    padding=(1, 1),
+                    channels=10,
+                    kernel_size=(3, 3),
+                    dilation=(3, 3))
+
+
+def test_ops():
+    run_ops('float32', 'custom[posites2]8', rtol=1, atol=1)
+    run_ops('float32', 'custom[posites2]16', rtol=0.01, atol=1)
+    run_ops('float32', 'custom[posites2]32')
+
+
+def test_conv2d():
+    # TODO(@gussmith23) slow and broken, needing refactor!
+    # run_conv2d('float32', 'custom[posit32]32')
+    pass
+
+
+def test_models():
+    # Expected posit8 might be faster, but it's not.
+    # run_model(get_mobilenet, (3, 224, 224), 'float32', 'custom[posit8]8')
+    # run_model(get_mobilenet, (3, 224, 224), 'float32', 'custom[posit32]32')
+    # run_model(get_inception, (3, 299, 299), 'float32', 'custom[posit32]32')
+    # run_model(get_resnet, (3, 224, 224), 'float32', 'custom[posit32]32')
+
+    # Run cifar-10 sizes to be a little faster...
+    run_model(get_mobilenet, (3, 32, 32),
+              'float32',
+              'custom[posites2]32',
+              num_classes=10)
+    # run_model(get_inception, (3, 32, 32),
+    #           'float32',
+    #           'custom[posites2]32',
+    #           num_classes=10)
+    # run_model(get_resnet, (3, 32, 32),
```

Comment:
resnet (x is posits, y is floats): 
```Mismatched elements: 10 / 10 (100%)
Max absolute difference: 0.01120126
Max relative difference: 0.10072963
 x: array([[0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1]], dtype=float32)
 y: array([[0.097013, 0.097102, 0.105917, 0.099155, 0.095107, 0.096329,
        0.097079, 0.111201, 0.102866, 0.09823 ]], dtype=float32)
```

@gussmith23  should we simply increase atol/rtol or look more into why posits are doing so poorly?

Reply:
This isn't right! Something's going wrong. This sounds dumb, but I'm going to just try and apply SimplifyInference and see if it fixes things...

Reply:
We should debug small Resnet. Something is going wrong. The easiest way to debug is to return early from the Resnet constructor:

``` diff
modified   python/tvm/relay/testing/resnet.py
@@ -162,6 +162,7 @@ def resnet(units,
     data = relay.var("data", shape=data_shape, dtype=dtype)
     data = layers.batch_norm_infer(data=data, epsilon=2e-5, scale=False, name='bn_data')
     (_, _, height, _) = data_shape
+    return relay.Function(relay.analysis.free_vars(data), data)
     if height <= 32:            # such as cifar10
         body = layers.conv2d(
             data=data, channels=filter_list[0], kernel_size=(3, 3),
@@ -194,7 +195,6 @@ def resnet(units,
     flat = relay.nn.batch_flatten(data=pool1)
     fc1 = layers.dense_add_bias(data=flat, units=num_classes, name='fc1')
     net = relay.nn.softmax(data=fc1)
-    return relay.Function(relay.analysis.free_vars(net), net)
 
 
 def get_net(batch_size,
```
This should work well with the current test file.
You can move the early return up and down to see where things diverge.
In my limited testing, it looks like posits are going to NaN after this first batch norm! Running the above version of resnet produces NaNs, but if we move the early return one layer up, then things are fine:

``` diff
modified   python/tvm/relay/testing/resnet.py
@@ -160,6 +160,7 @@ def resnet(units,
     num_unit = len(units)
     assert num_unit == num_stages
     data = relay.var("data", shape=data_shape, dtype=dtype)
+    return relay.Function(relay.analysis.free_vars(data), data)
     data = layers.batch_norm_infer(data=data, epsilon=2e-5, scale=False, name='bn_data')
     (_, _, height, _) = data_shape
     if height <= 32:            # such as cifar10
@@ -194,7 +195,6 @@ def resnet(units,
     flat = relay.nn.batch_flatten(data=pool1)
     fc1 = layers.dense_add_bias(data=flat, units=num_classes, name='fc1')
     net = relay.nn.softmax(data=fc1)
-    return relay.Function(relay.analysis.free_vars(net), net)
 
 
 def get_net(batch_size,
```

This version of Resnet literally doesn't do anything, and just returns the input. But we know it works, and we can see that things go wrong after that first batch norm.

My first thought was "oh, we need to run `SimplifyInference`!" So I did:
``` diff
modified   tests/python/unittest/test_custom_datatypes.py
@@ -49,6 +49,8 @@ def change_dtype(src, dst, module, params):
     return module, params
 
 def compare(module, input, src_dtype, dst_dtype, rtol, atol, params = {}):
+    module = relay.transform.SimplifyInference()(module)
+
     ex = relay.create_executor("graph", mod=module)
 
     correct = ex.evaluate()(*input, **params)
```

This fixes the batch norm...somewhat. We still get numerical errors between the outputs of the batch norms:

AssertionError: 
Not equal to tolerance rtol=0.0001, atol=0.0001
```
Mismatch: 99.2%
Max absolute difference: 0.01599717
Max relative difference: 0.01600081
 x: array([[[[0.692229, 0.071629, 0.014329, ..., 0.984048, 0.871523,
          0.445013],
         [0.728206, 0.627117, 0.563761, ..., 0.378826, 0.326279,...
 y: array([[[[0.681328, 0.070501, 0.014103, ..., 0.968551, 0.857798,
          0.438005],
         [0.716737, 0.61724 , 0.554882, ..., 0.37286 , 0.321141,...
```

There shouldn't be this amount of numerical error between posit32es2 and float32 -- posit32es2 should be far more accurate than float32 at values around [0,1]. 

Here's a debugging list:
- I think we do still need to run `SimplifyInference` explicitly. This is confusing, because it's run in `Optimize()`...maybe it's not getting run?
  + [ ] Add `SimplifyInference` somewhere (maybe in `compare()` like I did? I think that might be the best place)
  + [ ] Figure out whether `SimplifyInference` is seemingly not getting run later on
- [ ] I think we should add a test for batch norm itself, because it's been such a thorn in the side of this project. Have a test that runs just a batch norm through your `compare()`. Once we fix batch norm, hopefully everything else will be fixed.
- To debug the numerical issues with batch norm, here are a few dead-simple ideas off the top of my head:
  + [ ] Make sure that the workload is using the posit32 C functions. Maybe your `create_lower_func()` call is lowering them to the wrong bitwidth?
  + [ ] Make sure the params going into the batch norm are the same between float32 and posit32. (Batch norms take four parameters: mean, var, beta, gamma)

Let me know where you get with this! Good luck!